### PR TITLE
Default trainer

### PR DESCRIPTION
**Pull request by samedii**
I made a few changes to the default trainer that we can discuss around here. I think there are more things to consider and I don't want to change too much without getting @FelixAbrahamsson input on what the pitfalls are.

Changes made:
- Renamed flag to `n_batches_per_step` - I believe this is a better name since in my opinion `step` is from `optimizer.step` and when we are accumulating gradients from batches we are not actually making any steps.
- Removed mixup from the default trainer
- Kept model device solution but only getting the device when creating the trainer
- Added two handlers to the default trainer

Things to consider:
- Should we and can we just pass the `batch` object rather than forcing a dict with keys `features` and `targets`?
- Should we move out some or all of the `process_batch` function outside of the trainer? The trainer would not have to know about the model anymore nor the criterion.
- The trainer and evaluator are tightly coupled when attaching handlers which makes it difficult to move in some things into the default trainer like I first thought

@FelixAbrahamsson  @Jim-Holmstroem  What are your thoughts?

### CODE REVIEW

**Comment by samedii** [comment]
Fixed conflicts with recent merges

**Review by FelixAbrahamsson — COMMENTED** [review]
Looks good!

**Review by Jim-Holmstroem — APPROVED** [review]

**Review by FelixAbrahamsson — APPROVED** [review]
